service/worker/dlq/workflow.go

```diff
@@ -0,0 +1,169 @@
+	errorTypeInvalidRequest      = "dlq-error-type-invalid-request"
+	deleteTasksActivityName      = "dlq-delete-tasks-activity"
+
+	deleteTasksTimeout = 15 * time.Second
+)
+
+// Module provides a [workercommon.WorkerComponent] annotated with [workercommon.WorkerComponentTag] to the graph, given
+// a [HistoryServiceClient] dependency.
+var Module = workercommon.AnnotateWorkerComponentProvider(newComponent)
+
+func newComponent(client HistoryServiceClient) workercommon.WorkerComponent {
+	return &workerComponent{
+		client: client,
+	}
+}
+
+//revive:disable:import-shadowing this doesn't actually shadow imports because it's a method, not a function
+func (c *workerComponent) workflow(ctx workflow.Context, params WorkflowParams) error {
+	if params.WorkflowType == WorkflowTypeDelete {
+		ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			StartToCloseTimeout: deleteTasksTimeout,
+			TaskQueue:           primitives.DLQActivityTQ,
+		})
+		future := workflow.ExecuteActivity(ctx, deleteTasksActivityName, params.DeleteParams)
+
+		return future.Get(ctx, nil)
+	}
+
+	return temporal.NewNonRetryableApplicationError(params.WorkflowType, errorTypeInvalidWorkflowType, nil)
+}
+
+func (c *workerComponent) deleteTasks(ctx context.Context, params DeleteParams) error {
+	req := &historyservice.DeleteDLQTasksRequest{
+		DlqKey: &commonspb.HistoryDLQKey{
+			TaskCategory:  int32(params.TaskCategory),
+			SourceCluster: params.SourceCluster,
+			TargetCluster: params.TargetCluster,
+		},
+		InclusiveMaxTaskMetadata: &commonspb.HistoryDLQTaskMetadata{
+			MessageId: params.MaxMessageID,
+		},
+	}
+
+	_, err := c.client.DeleteDLQTasks(ctx, req)
+	if err != nil {
+		// Don't retry invalid-argument or not-found errors.
+		code := serviceerror.ToStatus(err).Code()
+		if code == codes.InvalidArgument || code == codes.NotFound {
+			return temporal.NewNonRetryableApplicationError(
+				"invalid argument deleting DLQ tasks", errorTypeInvalidRequest, err,
+			)
+		}
+		// Use the default retry policy for all other errors.
+		return err
+	}
+
+	return err
+}
+
+func (c *workerComponent) RegisterWorkflow(registry sdkworker.Registry) {
+	registry.RegisterWorkflowWithOptions(c.workflow, workflow.RegisterOptions{
+		Name: WorkflowName,
+	})
+}
+
+func (c *workerComponent) DedicatedWorkflowWorkerOptions() *workercommon.DedicatedWorkerOptions {
+	// use default worker
+	return nil
+}
+
+func (c *workerComponent) RegisterActivities(registry sdkworker.Registry) {
+	registry.RegisterActivityWithOptions(c.deleteTasks, activity.RegisterOptions{
+		Name: deleteTasksActivityName,
+	})
+}
+
+func (c *workerComponent) DedicatedActivityWorkerOptions() *workercommon.DedicatedWorkerOptions {
+	return &workercommon.DedicatedWorkerOptions{
+		TaskQueue: primitives.DLQActivityTQ,
+		Options: sdkworker.Options{
+			BackgroundActivityContext: headers.SetCallerType(
+				context.Background(),
+				headers.CallerTypeBackground,
```

Comment:
CallerTypePreemptable :)

Reply:
Thanks for catching that! Total brain fart